sakai-evaluation-impl/src/java/org/sakaiproject/evaluation/logic/externals/EvalExternalLogicImpl.java

```diff
@@ -686,35 +772,97 @@ public int countEvalGroupsForUser(String userId, String permission) {
-            // need to remove the admin user or else they show up in unwanted places
-            if (userIds.contains(ADMIN_USER_ID)) {
-                userIds.remove(ADMIN_USER_ID);
+    public Set<String> getUserIdsForEvalGroup( String evalGroupID, String permission, Boolean sectionAware )
+    {
+        // Determine the site ID (strip out the section prefix and ID if present)
+        String siteID = evalGroupID;
+        boolean isSectionGroupID = false;
+        if( siteID.contains( EvalConstants.GROUP_ID_SECTION_PREFIX ) )
+        {
+            isSectionGroupID = true;
+            siteID = siteID.substring( 0, siteID.indexOf( EvalConstants.GROUP_ID_SECTION_PREFIX ) );
+        }
+
+        // If it's not section aware...
+        Set<String> userIDs = new HashSet<>();
+        if( !sectionAware )
+        {
+            // Get the list normally
+            List<String> azGroups = new ArrayList<>();
+            azGroups.add( siteID );
+            userIDs.addAll( authzGroupService.getUsersIsAllowed( permission, azGroups ) );
+            if( userIDs.contains( ADMIN_USER_ID ) )
+            {
+                userIDs.remove( ADMIN_USER_ID );
             }
         }
 
-        return userIds;
-    }
+        // Otherwise, it's section aware but we only need to run the following if the sectin prefix is present in the evalGroupID
+        else if( isSectionGroupID )
+        {
+            // Determine the section ID
+            String sectionID = evalGroupID.substring( evalGroupID.indexOf( EvalConstants.GROUP_ID_SECTION_PREFIX ) + 9 );
+            if( siteID.contains( EvalConstants.GROUP_ID_SITE_PREFIX ) )
+            {
+                siteID = siteID.replace( EvalConstants.GROUP_ID_SITE_PREFIX, "" );
+            }
+
+            try
+            {
+                // Get all roles for the site
+                Site site = siteService.getSite( siteID );
+                Set<Role> siteRoles = site.getRoles();
+                Set<Role> siteRolesWithPerm = new HashSet<Role>();
+
+                // Determine which roles have the given permission
+                for( Role role : siteRoles )
+                {
+                    if( role.getAllowedFunctions().contains( permission ) )
+                    {
+                        siteRolesWithPerm.add( role );
+                    }
+                }
+
+                // Get the section and the user role map for the section
+                Section section = courseManagementService.getSection( sectionID );
+                Map<String, String> userRoleMap = sectionRoleResolver.getUserRoles( courseManagementService, section );
+
+                // Loop through the user role map; if the user's section role is in the list of site roles with the permission, add the user to the list
+                for( Entry<String, String> userRoleEntry : userRoleMap.entrySet() )
+                {
+                    for( Role role : siteRolesWithPerm )
```

Comment:
Maybe store the roleId in siteRolesWithPerm (as a list with a size hint since we know the max value) and then just call siteRolesWithPerm.contains(userRoleEntry.getValue()) rather than having an inner loop here.